### PR TITLE
IGNITE-12279: [ML] Added support for using H2O MOJO for model inference

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

**Comment by michalkurka** [comment]
@zaleslaw this is is a simple initial version of support for scoring H2O-3 MOJO models

**Comment by zaleslaw** [comment]
Thank you, for your contribution, will review next week.

чт, 10 окт. 2019 г., 20:49 Michal Kurka <notifications@github.com>:

File: examples/src/main/java/org/apache/ignite/examples/ml/h2o/H2OMojoModelParserExample.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.HashMap;
+import java.util.Scanner;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+/**
+ * This example demonstrates how to import H2O MOJO model and use imported model for distributed inference in Apache
+ * Ignite.
+ */
+public class H2OMojoModelParserExample {
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/h2o/H2OMojoModelParserExample.java]: Please, move this to the ml/inference/h2o folder

File: modules/ml/h2o-model-parser/pom.xml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.ignite</groupId>
+            <artifactId>ignite-ml</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>ai.h2o</groupId>
+            <artifactId>h2o-genmodel</artifactId>
+            <version>3.26.0.6</version>
```
Comment (zaleslaw) [inline modules/ml/h2o-model-parser/pom.xml]: Please, define as separate property at the start of the file.
have a look at the ignite-ml-tensorflow-ml-parser with 

    <properties>
        <tensorflow.version>1.12.0</tensorflow.version>
    </properties>

**Comment by zaleslaw** [comment]
Dear @michalkurka  could you please fix minor notes from me?

**Comment by michalkurka** [comment]
@zaleslaw sure, on it - thanks

File: examples/src/main/java/org/apache/ignite/examples/ml/h2o/H2OMojoModelParserExample.java (hunk 1 above)
Comment (michalkurka) [reply examples/src/main/java/org/apache/ignite/examples/ml/h2o/H2OMojoModelParserExample.java]: Done, thanks for noticing

**Comment by michalkurka** [comment]
@zaleslaw thanks for the review - comments incorporated

File: modules/ml/h2o-model-parser/src/main/java/org/apache/ignite/ml/h2o/H2OMojoModelParser.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.ignite.ml.inference.parser.ModelParser;
+import org.apache.ignite.ml.math.primitives.vector.NamedVector;
+
+/**
+ * H2O MOJO model parser.
+ */
+public class H2OMojoModelParser implements ModelParser<NamedVector, Double, H2OMojoModel> {
+    /** */
+    private static final long serialVersionUID = -170352744966205716L;
+
```
Comment (zaleslaw) [inline modules/ml/h2o-model-parser/src/main/java/org/apache/ignite/ml/h2o/H2OMojoModelParser.java]: Remove empty line

File: modules/ml/h2o-model-parser/src/main/java/org/apache/ignite/ml/h2o/H2OMojoModelParser.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        for (int i = 0; i < mojoModel.nfeatures(); i++) {
+            if (mojoModel.getDomainValues(i) != null) {
+                String columnName = mojoModel.getNames()[i];
+                throw new UnsupportedOperationException("Unsupported MOJO model: only models using trained using " +
+                        "OneHotExplicit categorical encoding and models without categorical features are currently supported. " +
+                        "Column `" + columnName + "` is categorical.");
+            }
+        }
+    }
+
```
Comment (zaleslaw) [inline modules/ml/h2o-model-parser/src/main/java/org/apache/ignite/ml/h2o/H2OMojoModelParser.java]: Remove empty line

File: examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * Test model resource name.
+     */
+    private static final String MODEL_RES = "examples/src/main/resources/models/h2o/agaricus-gbm-mojo.zip";
+
+    /**
+     * Test data.
+     */
+    private static final String DATA_RES = "examples/src/main/resources/datasets/agaricus-test-data.txt";
+
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java]: Remove empty line

File: examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.examples.ml.inference.h2o;
+
+import org.apache.ignite.Ignite;
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java]: Reorder imports, please, java imports should be first, after that should be added ignite imports like below, for exxample, also all unused imports should be removed

import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.Spliterator;
import java.util.function.Consumer;
import org.apache.ignite.lang.IgniteUuid;
import org.apache.ignite.ml.math.Blas;
import org.apache.ignite.ml.math.exceptions.math.CardinalityException;
import org.apache.ignite.ml.math.exceptions.math.ColumnIndexException;
import org.apache.ignite.ml.math.exceptions.math.RowIndexException;
import org.apache.ignite.ml.math.functions.Functions;
import org.apache.ignite.ml.math.functions.IgniteBiFunction;
import org.apache.ignite.ml.math.functions.IgniteDoubleFunction;
import org.apache.ignite.ml.math.functions.IgniteFunction;
import org.apache.ignite.ml.math.functions.IgniteTriFunction;
import org.apache.ignite.ml.math.functions.IntIntToDoubleFunction;
import org.apache.ignite.ml.math.primitives.vector.Vector;
import org.apache.ignite.ml.math.primitives.vector.impl.DenseVector;
import org.apache.ignite.ml.math.primitives.vector.impl.VectorizedViewMatrix;

Comment (michalkurka) [reply examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java]: Thanks, what is the rule regarding 3rd party libs - I see both before and after in the current codebase. Does it matter?

File: examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java (hunk 5 above)
Comment (michalkurka) [reply examples/src/main/java/org/apache/ignite/examples/ml/inference/h2o/H2OMojoModelParserExample.java]: Will do

**Comment by zaleslaw** [comment]
@michalkurka Congrats you with the first contribution to Apache Ignite, hope you will add something in future!

**Comment by zaleslaw** [comment]
Write about your donation on apache-ignite dev-list http://apache-ignite-developers.2346864.n4.nabble.com/H2O-integration-td44234.html

**Pull request merged**